### PR TITLE
Proper wayland support

### PR DESCRIPTION
**Pull request by snakedye**
Please follow this template, if applicable.

## Description

This is a very early version of the Wayland backend made using gtk-layer-shell-rs.

## Usage

When adding a widget or anything else that affects the configuration,
please provide a minimal example configuration snippet showcasing how to use it and

## Additional Notes

Since the transition to GTK4 has been halted, the door for Wayland support is reopened.

I think it would be great to reopen the Wayland [issue ](https://github.com/elkowar/eww/issues/10) to discuss more about how this should be done.

## Checklist

Please make sure you can check all the boxes that apply to this PR.

- [x] All widgets I've added are correctly documented.
- [x] The documentation in the `docs/content/main` directory has been adjusted to reflect my changes.
- [x] I used `cargo fmt` to automatically format all code before committing

### CODE REVIEW

File: src/display_backend.rs [hunk 1]
```diff
@@ -1,11 +1,171 @@
...
+                // .get_xid() as u32;
+            let root_window_geometry = self.conn.get_geometry(self.root_window)?.reply()?;
+
+            // Initialising a layer shell surface
+            gtk_layer_shell::init_for_window(window);
+            // Set the layer where the layer shell surface will spawn
+            gtk_layer_shell::set_layer(window, self.layer);
+            // Anchoring the surface to an edge
+            self.set_anchor(strut_def, window);
+            // I don't like the way NumWithWidth is used to define margins
```
Comment (elkowar) [inline src/display_backend.rs:61]: It pretty much has to be though, as it's the only way to make it possible to set margins in percentage - which definitely should be possible

File: src/display_backend.rs [hunk 2]
```diff
@@ -1,11 +1,171 @@
...
+            self.set_margin(strut_def, window);
+
+            // From my understanding the Atom equivalent on Wayland is something like APP_ID 
+            // This absolutely change but in the meanwhile I'll keep it here to not forget
```
Comment (elkowar) [inline src/display_backend.rs:65]: confused

File: src/display_backend.rs [hunk 3]
```diff
@@ -1,11 +1,171 @@
...
+        let backend = WaylandBackend::new()?;
+        backend.reserve_space_for(window, monitor, strut_def)?;
+        Ok(())
+    }
+
+    struct WaylandBackend {
+        layer: gtk_layer_shell::Layer;
+        conn: RustConnection<DefaultStream>,
+        // I have no idea what this is. Might look later
+        // root_window: u32,
```
Comment (elkowar) [inline src/display_backend.rs:29]: X-specific stuff, this would obviously need to go in the wayland backend

File: src/display_backend.rs [hunk 4]
```diff
@@ -1,11 +1,171 @@
...
+                    PropMode::Replace,
+                    win_id,
+                    self.atoms._NET_WM_STRUT_PARTIAL,
+                    self.atoms.CARDINAL,
+                    32,
+                    12,
+                    &strut_list,
+                )?
+                .check()?;
+            self.conn.flush()?;
```
Comment (elkowar) [inline src/display_backend.rs:88]: more x stuff that can be yeeted from here

File: src/display_backend.rs [hunk 5]
```diff
@@ -1,11 +1,171 @@
...
+            _NET_WM_STRUT,
+            _NET_WM_STRUT_PARTIAL,
+            WM_NAME,
+            UTF8_STRING,
+            COMPOUND_TEXT,
+            CARDINAL,
+            ATOM,
+            WM_CLASS,
+            STRING,
+        }
```
Comment (elkowar) [inline src/display_backend.rs:168]: Xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx stuff

File: src/display_backend.rs [hunk 6]
```diff
@@ -1,11 +1,171 @@
...
+            }
+            if margin_right>0 {
+                gtk_layer_shell::set_margin(window, gtk_layer_shell::Edge::Right, margin_right);
+            }
+            if margin_bottom>0 {
+                gtk_layer_shell::set_margin(window, gtk_layer_shell::Edge::Bottom, margin_bottom);
+            }
+            if margin_left>0 {
+                gtk_layer_shell::set_margin(window, gtk_layer_shell::Edge::Left, margin_left);
+            }
```
Comment (elkowar) [inline src/display_backend.rs:145]: why the if-checks? shouldn't this, in case you run it twice, once with a value > 0, then with 0, reset the margins of that side to 0?

Comment (snakedye) [reply src/display_backend.rs:145]: I didn't took into consideration theses values could be edited after run time. In this case the if aren't needed. If was just to not pointlessly set 0 margins.

File: src/display_backend.rs (hunk 1 above)
Comment (snakedye) [reply src/display_backend.rs:61]: Fair.

File: src/display_backend.rs (hunk 2 above)
Comment (snakedye) [reply src/display_backend.rs:65]: Verbalized internal dialog. I speculated there could a wayland alternative to AtomCollection in x11rb. The comment is just a reminder to look this up eventually.

File: src/display_backend.rs (hunk 1 above)
Comment (snakedye) [reply src/display_backend.rs:61]: Since percentage works with the output dimension, I think I'll need to use the output management protocol. Probably with [smithay](https://docs.rs/smithay-client-toolkit/0.12.2/smithay_client_toolkit/).

**Comment by snakedye** [comment]
I can't import `gtk-layer-shell-rs` when I compile eww. It's not like I think it would compile anyway if the crate was added but I could at least start debugging the issues.

**Comment by snakedye** [comment]
It's fixed I had to pull gdk and gtk from git

**Comment by undefinedDarkness** [comment]
This works now which is pretty pog

**Comment by snakedye** [comment]
I think all that's left now is editing the documentation.

**Comment by undefinedDarkness** [comment]
Is it plug & play with X11 config?

**Comment by snakedye** [comment]
No there are extra config options for wayland

**Comment by undefinedDarkness** [comment]
:confused: if it can do extra stuff why cant it just work with X11 config, the extra stuff can be available too

**Comment by undefinedDarkness** [comment]
> It cant, X is shit and offers less flexibility
Pity but it makes sense

File: src/config/window_definition.rs [hunk 7]
```diff
@@ -52,6 +56,7 @@ impl RawEwwWindowDefinition {
         let stacking: WindowStacking = xml.parse_optional_attr("stacking")?.unwrap_or_default();
 
         // TODO maybe rename this to monitor?
+        // Yes, you should
```
Comment (elkowar) [inline src/config/window_definition.rs:59]: lmao 
```suggestion
```

File: src/config/window_definition.rs [hunk 8]
```diff
@@ -94,20 +105,29 @@ impl std::str::FromStr for Side {
...
             "b" | "bottom" => Ok(Side::Bottom),
+            "c" | "center" => Ok(Side::Center),
+            "tl" | "top-left" => Ok(Side::TopLeft),
+            "tr" | "top-right" => Ok(Side::TopRight),
+            "bl" | "bottom-left" => Ok(Side::BottomLeft),
+            "br" | "bottom-right" => Ok(Side::BottomRight),
             _ => Err(anyhow!(
-                "Failed to parse {} as valid side. Must be one of \"left\", \"right\", \"top\", \"bottom\"",
+                "Failed to parse {} as valid side. Must be one of \"left\", \"right\", \"top\", \"bottom\", \"top-right\", \
+                 \"top-left\", \"bottom-left\", \"bottom-right\"",
```
Comment (elkowar) [inline src/config/window_definition.rs:115]: You can use a raw string thingy here to get rid of all of the escaping - do `r#"Failed to parse ...... "left", "right", ......."#`, that get's rid of all of the `\"`

File: src/opts.rs [hunk 9]
```diff
@@ -7,6 +7,7 @@ use crate::{
     config::{AnchorPoint, WindowName},
     value::{Coords, PrimitiveValue, VarName},
 };
+// use crate::value::coords;
```
Comment (elkowar) [inline src/opts.rs:10]:
```suggestion

```

File: src/app.rs [hunk 10]
```diff
@@ -331,12 +330,28 @@ fn initialize_window(
...
+        // Sets the monitor where the surface is shown
+        match window_def.screen_number {
+            Some(index)=>{
+                let monitor = get_monitor(index);
+                gtk_layer_shell::set_monitor(&window, &monitor);
+            },
+            None=>{}
+        };
+        gtk_layer_shell::set_keyboard_interactivity(&window, window_def.focusable);
+    }
```
Comment (elkowar) [inline src/app.rs:346]: formatting REEEEEEEEEEEEEEEEEEEEEEEEEEEE
But also, can we put this into a function somehow? Just having a big cfg(feature...) block in the middle of the function is kinda ehhhhhhhhhh. a `wayland_initialize_window` thing or sth may be reasonable

File: src/config/window_definition.rs [hunk 11]
```diff
@@ -117,11 +137,38 @@ impl StrutDefinition {
     }
 }
 
+// Surface definition if the backend for Wayland is enable
+#[cfg(feature = "wayland")]
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Default)]
+pub struct StrutDefinition {
+    pub layer: WindowStacking,
```
Comment (elkowar) [inline src/config/window_definition.rs:144]: This is rather confusing, as the stacking is otherwise defined on the EwwWindowDefinition,.... it should probably stay there.

File: examples/eww-power/eww.xml [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+  <windows>
+    <window name="power-menu" screen="1">
+      <geometry x="0%" y="0%" width="600" height="10%"/> <!--Specifies geometry-->
+      <reserve side="center" x="0%" y="0%" stacking="fg" exclusive="false"/>
+      <widget>
+        <power-menu/> <!--This is the def we created earlier-->
+      </widget>
+    </window>
+  </windows>
+</eww>
```
Comment (elkowar) [inline examples/eww-power/eww.xml:52]: 🤔 if anything, another example should prolly go into a separate PR, no?

File: Cargo.toml [hunk 13]
```diff
@@ -10,22 +10,23 @@ homepage = "https://github.com/elkowar/eww"
...
+no-x11-wayland = []
 
 [dependencies]
-gtk = { version = "0.9", features = [ "v3_16" ] }
-gdk = { version = "", features = ["v3_16"] }
-gio = { version = "", features = ["v2_44"] }
+gtk = { git = "https://github.com/gtk-rs/gtk-rs" }
+gdk = { git = "https://github.com/gtk-rs/gtk-rs", features = ["v3_22"] }
+gdk-sys = { git = "https://github.com/gtk-rs/gtk-rs" }
+gio = { git = "https://github.com/gtk-rs/gtk-rs" }
```
Comment (elkowar) [inline Cargo.toml:22]: are the git deps here really necessary still? 😢

**Review by elkowar — CHANGES_REQUESTED** [review]
Epic! Still some things left tho, sorry for taking so long ^^

File: src/display_backend.rs [hunk 14]
```diff
@@ -1,11 +1,171 @@
...
+                    bottom=true;
+                }
+            }
+
+            // Anchors are if the window is pinned to each edge of the output
+            gtk_layer_shell::set_anchor(self.window, gtk_layer_shell::Edge::Left, top);
+            gtk_layer_shell::set_anchor(self.window, gtk_layer_shell::Edge::Right, right);
+            gtk_layer_shell::set_anchor(self.window, gtk_layer_shell::Edge::Top, top);
+            gtk_layer_shell::set_anchor(self.window, gtk_layer_shell::Edge::Bottom, bottom);
+        }
```
[the remark this answers is not in the document]
Comment (snakedye) [reply src/display_backend.rs]: They should be named anchors but I didn't want to touch too much stuff on the the X11 backed so I kept Side. Top_Left will anchor a surface on the top and left edge which will create a new anchor point in the left left edge of the screen.

> how do the new Top_left etc edges behave on X?

I think I'll just yeet them away from the X config so eww errors out when they are parsed.

File: examples/eww-power/eww.xml (hunk 12 above)
Comment (snakedye) [reply examples/eww-power/eww.xml:52]: I think this fits with the updated documentation. Having an appropriate example is a bonus.

File: Cargo.toml (hunk 13 above)
Comment (snakedye) [reply Cargo.toml:22]: I think so because gtk-layer-shell-rs only version is git (it's not on crates.io)  so it is most likely based on the latest versions of gtk/gdk libs.

File: src/config/window_definition.rs [hunk 15]
```diff
@@ -84,6 +84,11 @@ pub enum Side {
     Left,
     Right,
     Bottom,
+    Center,
+    Top_Left,
+    Top_Right,
+    Bottom_Left,
+    Bottom_Right,
```
[the remark this answers is not in the document]
Comment (snakedye) [reply src/config/window_definition.rs]: I think it is doable with the strut_list.

File: Cargo.toml (hunk 13 above)
Comment (Th3Whit3Wolf) [reply Cargo.toml:22]: If you specify the rev it may save you some pain in the future in case somethings breaks between commits.

Comment (elkowar) [reply Cargo.toml:22]: ^ in this case, we should definitely specify the git revision explicitly!  @snakedye

File: examples/eww-power/eww.xml (hunk 12 above)
Comment (elkowar) [reply examples/eww-power/eww.xml:52]: Ehhhhh,....
Also, regardless of if this is fitting or necessary, an example should definitely not hardcode paths into the system installation for symbols like that :/

**Comment by elkowar** [comment]
A couple more things:
- The default feature-set should still be `x11`, as that's what the large majority uses
image

- Currently, even after setting the value back to x11, it doesn't compile (meaning that there are multiple errors in the x11 variant of the code that must be fixed)
image

**Comment by snakedye** [comment]
The code compiles with regular git deps but the revs are driving me insane. Maybe I have no idea what revs are idk, this has been slowing me down for a week and I don't know what more I can/should do. Pls help

**Comment by Th3Whit3Wolf** [comment]
@snakedye  

What happends if you replace `rev="abd4ce6f"` with `rev = "abd4ce6f6122c39e80fec457a29e8ad1aa063b81"`

and `rev="bf097e83"` with `rev = "f7c920c478307b6b7cba612585662d71fd750a90"`

**Comment by Th3Whit3Wolf** [comment]
@snakedye I'm having some issues too. If you remove the revision on all but gtk-layer-shell-rs it builds. I'm not sure where the build errors stem from.

**Comment by Th3Whit3Wolf** [comment]
@snakedye @elkowar This is almost certainly a dumb question but would it be possible to use [gtk-layer-shell](https://crates.io/crates/gtk-layer-shell) from crates.io? [Gtk-layer-shell-rs's author](https://github.com/subgraph/gtk-layer-shell-rs/issues/3) decided not to publish his project because of the similar crates.

**Comment by snakedye** [comment]
@Th3Whit3Wolf It fails to compile and breaks some gdk stuff.
image

**Comment by Th3Whit3Wolf** [comment]
> @Th3Whit3Wolf It fails to compile and breaks some gdk stuff.
> image

Do you not get these errors when not specifying the commits?

**Comment by snakedye** [comment]
To use gtk-layer-shell crate I changed every git deps for the one on crates.io.
image

**Comment by Th3Whit3Wolf** [comment]
> @Th3Whit3Wolf To use gtk-layer-shell crate I changed every git deps for the one on crates.io.
> image

Sounds a lot better for reproducibility!

**Comment by snakedye** [comment]
Wayland compiles and run fine, X11 doesn't because of something in the backend. So @elkowar now is your time to shine.

**Comment by snakedye** [comment]
Now everything works :pogphisch:

**Comment by elkowar** [comment]
@snakedye resolve conflicts plox uwu

**Comment by elkowar** [comment]
Sorry that this is going so slowly ^^' life is stressful rn

**Comment by elkowar** [comment]
image
:(

**Comment by snakedye** [comment]
@snakedye I tried to solve as many conflicts as I could. I doubt I can resolve the Cargo.lock, the toml is just being unreasonably picky about a Return and the one in app.rs is due either `cargo fmt` or I can't change the `use std::{collections::HashMap, path::PathBuf};` without breaking something.

**Comment by elkowar** [comment]
I mean, worst case, just delete cargo.lock, have it regenerate, and then use that. the toml and app.rs conflicts should be easily resolvable in a rebase 👀

**Comment by snakedye** [comment]
How can I rebase from/to your master branch?

**Comment by elkowar** [comment]
You set up my repo as another remote (git remote add upstream https://blaboabla) then you fetch that and git rebase upstream/master

**Comment by snakedye** [comment]
I legitimately feel like a monke.

**Comment by snakedye** [comment]
Epic.

**Comment by elkowar** [comment]
https://github.com/elkowar/eww/pull/125/files#diff-1f7271b5c39d126a92941ed7e215373e905ec0bd05cec17e4227480cd9ae96fd

What even is this file, and why is it there 🤔

**Review by elkowar — CHANGES_REQUESTED** [review]
Some more weird stuff here,..... Probably happened while merging stuff. Did this even compile on wayland? :thinking:

**Comment by snakedye** [comment]
> Some more weird stuff here,..... Probably happened while merging stuff. Did this even compile on wayland? thinking

No but the build error I get is this.
image

which seems related to some experimental feature I don't have although I have rustup installed.

**Comment by elkowar** [comment]
run a `rustup update`, you're probably on an outdated version of the nightly compiler

**Comment by snakedye** [comment]
The wayland and x11 backend compile. The wayland backend works fine for me.

**Comment by elkowar** [comment]
The options currently still seem fucked:
- `<reserve>` requires an x, y, and exclusive option (which it definitely shouldn't require, if those are necessary at all,....)
- `<geometry>`'s positioning doesn't work, whereas the window _can_ be positioned using reserve,.... :thonk:

**Comment by snakedye** [comment]
exclusive is basically reserve. I don't really think there's a need to have reserve then.

**Comment by elkowar** [comment]
Wasn't part of the wayland ide that you can reserve space in the middle of the screen? If so, x and y would be required, no? 

I dont really get what exclusive=false would imply tho. 

Also, main issue is definitely that x and y on geometry are broken. Also, the docs should explain stuff a bit more clearly imo, and, if you're extending the example config anyways, make sure to include all options that one could set in there.

**Comment by snakedye** [comment]
> Wasn't part of the wayland ide that you can reserve space in the middle of the screen? If so, x and y would be required, no?

It's not the behavior recommended by layershell.

>I dont really get what exclusive=false would imply tho.

Exclusive just tells the compositor if a surface and its region can be occupied by another. If set to false, other surfaces can be drawn over Eww if not the surface cannot be occupied by another. This is generally translated by a shift in the position of windows to accommodate the surface.

>Also, main issue is definitely that x and y on geometry are broken. Also, the docs should explain stuff a bit more clearly imo, and, if you're extending the example config anyways, make sure to include all options that one could set in there.

I'll work on that.

**Comment by elkowar** [comment]
> Exclusive just tells the compositor if a surface and its region can be occupied by another. If set to false, other surfaces can be drawn over Eww if not the surface cannot be occupied by another. This is generally translated by a shift in the position of windows to accommodate the surface.

So if I set exclusive to false, does the reserve still do _anything_? or is it just a noop then

**Comment by snakedye** [comment]
it could still be used in the config to set the stacking and side but the `reserve_space` function in eww would literally only serve to toggle exclusive.

**Comment by snakedye** [comment]
just looked what changed and yeah reserve is useless. I will promptly yeet it

**Comment by elkowar** [comment]
> just looked what changed and yeah reserve is useless. I will promptly yeet it

Is how is `<reserve>` suddenly useless? Like, reserving space is still necesary, no? It just doesn't make sense to have reserve but then also ahve `exclusive=false`..
Or how would I otherwise make it reserve space on screen 🤔

**Comment by snakedye** [comment]
To reserve space you would just need to have `exclusive=true`. I could leave `<reserve/>` but it would only have "exclusive" as a field.

**Comment by elkowar** [comment]
How would it know _where_ to reserve space?
Specifically on X, where the reservation doesn't just map to window position directly

**Comment by snakedye** [comment]
It will reserve the region the widget occupies so for example if you have a bar at the top of your screen, it will reserve the height of the bar from where it is anchored ie the top.

**Comment by elkowar** [comment]
Does it do that on X? how does it behave if the window is in the middle of the screen?

**Comment by snakedye** [comment]
That's left to compositor's implementation.

Here](https://github.com/swaywm/wlr-protocols/blob/d1598e82240d6e8ca57729495a94d4e11d222033/unstable/wlr-layer-shell-unstable-v1.xml#L152)'s the documentation.

**Comment by elkowar** [comment]
> > Does it do that on X?
> 
> It's specific to the Wayland backend.
> 
> > how does it behave if the window is in the middle of the screen?
> 
> That's left to compositor's implementation.
> 
> [Here](https://github.com/swaywm/wlr-protocols/blob/d1598e82240d6e8ca57729495a94d4e11d222033/unstable/wlr-layer-shell-unstable-v1.xml#L152)'s the documentation.

I mean, yes, my point is - `<reserve>` is also being used by the X backend for reserving space, so don't just yeet it :P
The remaining question is if there is any value in having reservation of space be separate from positioning on wayland. I'd argue, for the sake of consistency, keeping reserve makes sense,... no? I'm not sure on that one

**Comment by snakedye** [comment]
> I mean, yes, my point is - <reserve> is also being used by the X backend for reserving space, so don't just yeet it :P

Yeah it's still there. Poor phrasing on my end. I just meant making the whole block not necessary.

>The remaining question is if there is any value in having reservation of space be separate from positioning on wayland. I'd argue, for the sake of consistency, keeping reserve makes sense,... no? I'm not sure on that one

It makes sense but that block would only have that "exclusive" as a field. If you're fine with that it's alright. I don't think it changes a lot of things.

**Comment by elkowar** [comment]
I mean, does wayland automatically sync the `reserve` stuff with window position, or is that still manual? because the other option would be keeping position and size of window and position and the reserve offsets separate

**Comment by snakedye** [comment]
It does sync.

**Comment by elkowar** [comment]
always?

**Comment by snakedye** [comment]
always

**Comment by elkowar** [comment]
well then, yea, having the `exclusive` option _somewhere_ is still necessary - but yea, on wayland, reserve can then be ignored. that should be sufficiently documented tho

**Comment by snakedye** [comment]
Hopefully this should be enough.

File: docs/content/main/configuration.md [hunk 16]
```diff
@@ -222,17 +222,37 @@ The `<windows>` config should look something like this:
...
 - `focusable`: whether the window should be focusable by the windowmanager.
   This is necessary for things like text-input-fields to work properly.
   Possible values: `"true"`, `"false"`. Default: `"false"`
 - `screen`: Specifies on which display to show the window in a multi-monitor setup.
   This can be any number, representing the index of your monitor.
+- `exclusive`: Specifies whether or not a surface can be occupied by another. 
+  A surface can be a window, an Eww widget or any layershell surface.
+  The details on how it is actually implemented are left to the compositor.
+  This option is only valid on Wayland.
+  Possible values: `"true"`, `"false"`. Default: `"false"`
```
Comment (elkowar) [inline docs/content/main/configuration.md:258]: This is currently not configurable on thie window, but on the `<reserve>` tag.
IMO, it _should_ however actually be on the window, so fix code to have `exclusive` be on window on wayland (if reasonably clean to implement.)

File: Cargo.toml [hunk 17]
```diff
@@ -10,10 +10,12 @@ homepage = "https://github.com/elkowar/eww"
 
 
 [features]
-default = ["x11"]
+default = ["wayland"]
```
Comment (elkowar) [inline Cargo.toml:13]: No.

Default should still be x11, as _most_ people are on X11.

**Comment by elkowar** [comment]
Also still not a fan of the inclusion of another (very similar) example. It adds nothing, Is rather weird, and contains multiple things very specific to your setup. I'd yeet those.

File: Cargo.toml (hunk 17 above)
Comment (snakedye) [reply Cargo.toml]: I'll one day learn how to use flags.

Comment (elkowar) [reply Cargo.toml]: --no-default-features --features=wayland

**Comment by elkowar** [comment]
TODAY IS THE DAY
IT IS THE MOMENT!
IT IS TIME!

WAYLAND SUPPORT
IS GETTING
FUCKING 
MERGED!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
image

**Comment by buffet** [comment]
Proper more like improper, amirite?

**Comment by snakedye** [comment]
cry
